src/librustc/traits/project.rs

```diff
@@ -1247,6 +1247,34 @@ fn confirm_callable_candidate<'cx, 'gcx, 'tcx>(
                                               fn_sig,
                                               flag);
 
+    if selcx.projection_mode().is_any() {
```

Comment:
I'd like to try to make more progress on this, but I also want to make sure I understand the conversation here.

In particular, @nikomatsakis, when you wrote above that you expected a filter to be attached to the `FulfillmentContext`, did you mean that, or did you mean `SelectionContext`?

(I ask because I can at least see how a `SelectionContext` would connect up to the code as written, but attaching it to the `FulfillmentContext` seems like ... its leaving out details about how that filter would actually prevent this particular confirmation while not messing things up elsewhere. I have no intuition right now as to where the filter fundamentally *belongs*)